docs/proposals/20210713-resourcedistribution.md

```diff
@@ -0,0 +1,145 @@
+---
+title: Resource Distribution
+
+authors:
+  - "@veophi"
+
+reviewers:
+  - "@Fei-Guo"
+  - "@furykerry"
+  - "@FillZpp"
+
+creation-date: 2021-07-14
+last-updated: 2021-07-14
+status: implementable
+---
+
+# Resource Distribution
+- Provide a way to distribute some namespaced resource (e.g., `Secret`, `ConfigMap`) to other namespaces.
+
+## Table of Contents
+A table of contents is helpful for quickly jumping to sections of a proposal and for highlighting
+any additional information provided beyond the standard proposal template.
+[Tools for generating](https://github.com/ekalinin/github-markdown-toc) a table of contents from markdown are available.
+
+- [Resource Distribution](#resource-distribution)
+  - [Table of Contents](#table-of-contents)
+  - [Motivation](#motivation)
+  - [Proposal](#proposal)
+    - [User Stories](#user-stories)
+      - [Story 1: distribute to other namespaces](#story-1-distribute-to-other-namespaces)
+      - [Story 2: distribute to the whole cluster](#story-2-distribute-to-the-whole-cluster)
+    - [Implementation](#implementation)
+      - [WebHook Validation](#webhook-validation)
+      - [Secret Controller](#secret-controller)
+      - [Namespace Controller](#namespace-controller)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Implementation History](#implementation-history)
+
+## Motivation
+Sometimes, we want to apply some namespaced resources, such as `Secret` and `ConfigMap`,  to other namespaces, even to the whole cluster.
+
+For example, a `SidecarSet` with `imagePullSecrets` field may be injected into different namespaces, we must make sure the `Secrets` exist in all of these namespaces.
+In this situation, we may need a resource distribution mechanism to help us do it more easily.
+
+## Proposal
+**Main idea**: Distribute and sync the resources by the user-defined annotation.
+
+Next, we will take the `Secret Distribution` as an example to illustrate our proposal.
+The `ConfigMap Distribution` is similar with it.
+
+### User Stories
+#### Story 1: distribute to other namespaces
+```yaml
+# my-secret.yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: namespaced-secret
+  annotations:
+    openkruise.io/sync-to: "namespaces"
+    openkruise.io/sync-to/namespaces: "ns-1;ns-2;ns-3"
```

Comment:
1. openkruise.io/sync-to/namespaces -> e.g. openkruise.io/sync-to-namespaces
2. whether the user can configure only one annotations, e.g. openkruise.io/sync-to-namespaces
3. support all-namespaces, except kube-system